### PR TITLE
feat: use ipld-explorer-components@4.0.0

### PR DESCRIPTION
**Pull request by SgtPooki**
Note that this code is based on changes included in https://github.com/ipfs/ipld-explorer-components/pull/360 and must wait to be merged until we can publish a new ipld-explorer-components lib once that PR is merged. 

check out the preview: https://bafybeif65hxysjwygngoahx7v6a2ouzsfi4y24u7ovr4vd5ya4jhq7ngjy.on.fleek.co/

Check out the demo:

https://github.com/ipld/explore.ipld.io/assets/1173416/90802964-9e98-418c-9353-12270b1f71e1

fixes https://github.com/ipld/explore.ipld.io/issues/117

---

- [x] ~~Note that fleek build is failing due to being set to node16,~~ **RESOLVED**

image

- [x] ~~but we can't update fleek settings unless https://github.com/ipld/explore.ipld.io/pull/119/commits/5bb40dd991ce662739e50fc67db9769ec143cc83 is merged to master.~~ **RESOLVED** see https://github.com/ipld/explore.ipld.io/pull/120

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -28,43 +28,45 @@
...
-    "ipld": "^0.21.1",
-    "ipld-bitcoin": "^0.1.9",
-    "ipld-dag-cbor": "^0.13.1",
-    "ipld-dag-pb": "^0.15.2",
-    "ipld-ethereum": "^2.0.3",
-    "ipld-explorer-components": "^2.5.1",
-    "ipld-git": "^0.2.3",
-    "ipld-raw": "^2.0.1",
-    "ipld-zcash": "^0.4.1",
+    "ipld-explorer-components": "file:../../ipfs/ipld-explorer-components/ipld-explorer-components-3.0.3.tgz",
```
Comment (SgtPooki) [inline package.json:32]: will be updated once https://github.com/ipfs/ipld-explorer-components/pull/360 is merged

File: .env [hunk 2]
```diff
@@ -1 +1 @@
-REACT_APP_VERSION=$npm_package_version
+VITE_VERSION=$npm_package_version
```
Comment (SgtPooki) [inline .env:1]: part of vite migration

File: index.html [hunk 3]
```diff
@@ -8,17 +8,8 @@
...
-      Notice the use of %PUBLIC_URL% in the tags above.
-      It will be replaced with the URL of the `public` folder during the build.
-      Only files inside the `public` folder can be referenced from the HTML.
-
-      Unlike "/favicon.ico" or "favicon.ico", "%PUBLIC_URL%/favicon.ico" will
-      work correctly both with client-side routing and a non-root public URL.
-      Learn how to configure a non-root public URL by running `npm run build`.
-    -->
+    <link rel="manifest" href="/manifest.json">
+    <link rel="shortcut icon" href="/favicon.ico">
```
Comment (SgtPooki) [inline index.html:12]: part of vite migration

File: index.html [hunk 4]
```diff
@@ -36,5 +27,6 @@
       To begin the development, run `npm start` or `yarn start`.
       To create a production bundle, use `npm run build` or `yarn build`.
     -->
+    <script type="module" src="/src/index.jsx"></script>
```
Comment (SgtPooki) [inline index.html:30]: part of vite migration

File: package.json [hunk 5]
```diff
@@ -3,17 +3,17 @@
...
-    "test:serve": "ecstatic build --cache=0 --port=3001",
-    "test:e2e": "react-scripts test --roots test/ --testMatch '**/*.test.js'",
-    "test:e2e:ci": "URL=http://localhost:3001 run-p --race test:serve test:e2e",
-    "analyze": "webpack-bundle-analyzer build/bundle-stats.json",
+    "start": "vite",
+    "build": "vite build",
+    "serve": "vite preview",
+    "test": "run-s test:unit test:e2e",
+    "test:unit": "vitest run --environment=node",
+    "test:unit-playwright": "playwright test -c playwright-ct.config.js",
```
Comment (SgtPooki) [inline package.json:13]: not currently used, but available to run unit tests via playwright if we want to normalize on tests.

File: src/App.jsx [hunk 6]
```diff
@@ -11,7 +11,7 @@ export class App extends Component {
     doUpdateUrl: PropTypes.func.isRequired,
     queryObject: PropTypes.object.isRequired,
     registerServiceWorker: PropTypes.func,
-    route: PropTypes.oneOfType([PropTypes.func, PropTypes.element]).isRequired
+    route: PropTypes.oneOfType([PropTypes.func, PropTypes.element, PropTypes.elementType]).isRequired
```
Comment (SgtPooki) [inline src/App.jsx:14]: fixes an in-app error regarding prop types mismatch.

File: test/e2e/smoke.test.js [hunk 7]
```diff
@@ -1,29 +1,49 @@
...
-afterAll(() => browser.close())
+// beforeAll(async () => { browser = await launch() })
+// afterAll(() => browser.close())
 
-it('should resolve a dag-pb cid with a path', async () => {
-  // Save 3s per run by using the initial page rather than creating a new one!
-  // const page = await browser.newPage()
-  const page = (await browser.pages())[0]
+// it('should resolve a dag-pb cid with a path', async () => {
+//   // Save 3s per run by using the initial page rather than creating a new one!
```
Comment (SgtPooki) [inline test/e2e/smoke.test.js:11]: todo: need to migrate this test code to playwright.

**Comment by SgtPooki** [comment]
related issue found when doing js-ipfs deprecation: https://github.com/ipfs/js-ipfs/issues/1416

**Review by lidel — CHANGES_REQUESTED** [review]
Tested latest version from  https://bafybeiepdrogejw7eji6qdbq3hpd3ewuuhecyvf4nyofvhntsbyb3w7tqa.on.fleek.co

@SgtPooki looks ok for available content, but trying to fetch something new timeouts bit too early? Lmk if I got things right.

I tried to add a PDF and explore its CID (`bafybeibms6xfhs4kx3tun7dg6bjbj5obca7yj43cl5jl3pc2n7muw6ipfm`) and the gateway request for the root block  was stuck in "Pending":

> 
 
 
 and after a while that request got cancelled:
 
 >  


So two problems:
1. Why is request for single block in `Pending`? It should fhit gateway asap, before  anything else. Are WebTransport requests still suffocating the request queue browser allocates to the page?
2. Why it cancels (timeouts?) so fast? As long gateway request is in progress or pending, it should not timeout.

**Comment by SgtPooki** [comment]
> Why is request for single block in Pending? It should [hit] gateway asap, before anything else. Are WebTransport requests still suffocating the request queue browser allocates to the page?

Helia should not be making any requests except for to the delegatedClient via kubo-rpc-client: https://github.com/ipfs/ipld-explorer-components/blob/ca19310defb84611f55c46b1efd399217ba4a8c6/src/lib/init-helia.ts#L28-L63

> Why it cancels (timeouts?) so fast? As long gateway request is in progress or pending, it should not timeout.

Mostly to keep the site responsive and gateway-friendly. I've currently got the [timeout set to 30 seconds](https://github.com/ipfs/ipld-explorer-components/blob/ca19310defb84611f55c46b1efd399217ba4a8c6/src/lib/get-raw-block.ts#L108C98-L108C103). I've heard significant feedback from users that explore.ipld.io would spin and spin and they never knew if their content was going to load. I think a much better approach is to fail early, and have the users refresh/retry if they want. 

Were you able to fetch your content from http://ipfs.io/ipfs/bafybeibms6xfhs4kx3tun7dg6bjbj5obca7yj43cl5jl3pc2n7muw6ipfm directly? Maybe you removed the content, but I just attempted to access it and it failed after 

We shouldn't have users guessing whether a request is going to succeed or not. Also, for failed gateway requests, the gateway keeps the connection open for up to 2 minutes, and sends back almost 1K of data: 

image

If we cancel requests that are pending for 30 seconds, we can free up the gateway to do other work, and the user can re-request things if they want. If there is a different timeout value you think would be better or a better way to handle this without significant feature work, then please let me know.

---

#### Drawbacks

* The code is currently set to try the [gateways in serial](https://github.com/ipfs/ipld-explorer-components/blob/ca19310defb84611f55c46b1efd399217ba4a8c6/src/lib/get-raw-block.ts#L76) which means if ipfs.io takes more than `timeout` ms, dweb.link is never attempted. If ipfs.io takes less than `timeout` ms, then dweb.link is never attempted. These could be made parallel since we removed the non-delegated Helia fetch attempts. 

#### Other things to think about 

* Higher default timeout - Gives users the expectation that something will resolve, when in all likelihood, if it hasn't resolved in 30 seconds, it probably won't. In my experience improving ipld-explorer-components, 30 seconds was a pretty nice spot. 
* Better error message - I looked into displaying a better error message previously, but the way the network requests are hidden so deep in ipld-explorer-components, it was a larger task than I wanted to fit into https://github.com/ipfs/ipld-explorer-components/pull/360.
    * It would be great to be able to show more details about failed fetch attempts ("failed to load via ipfs.io: Gateway Timeout", "failed to load via dweb.link: Gateway Timeout", "failed to load via local node: request blocked", etc..) and also where we were able to successfully fetch the request from
* Configurable timeouts - Allow users to specify whether they want the request to timeout (i'm just curiously browsing), or whether they're willing to wait for long enough to load it (I need to actually explore this content for some reason)
* Stagger parallel gateway requests - We should not broadcast to all gateways at the same time, but we could delay the requests to subsequent gateways by something like`timeout/2` for example. 
* Separate timeouts + staggered requests - We could have a global-get-raw-block-timeout and stagger requests to gateways by (N+1 gateways tested / global-get-raw-block-timeout)

**Review by lidel — APPROVED** [review]
Thank you @SgtPooki, failing fast at 30s mark makes sense as a trade-off,  
the problem is bad UX on timeout – user adds data to their local node, then tries to inspect it, timeouts, app looks broken:

>  

One way to fix this without rewriting the entire thing is to inspect error before printing the red box and if it is timeout/promise cancellation,  replace it with friendly "Failed to fetch content in 30s, refresh page to retry or try different CID."?  It would be good enough to unblock this PR.

Since  ipld-explorer-components@4.0.0  is released, would it be ok to merge this as-is, and fill issues for any remaining work ("Drawbacks", "Other things to think about"), and do it  in future PRs (if we have time).

Would be shame if this PR got stuck in review forever: It is already better than the old version! :)

**Comment by SgtPooki** [comment]
> Failed to fetch content in 30s, refresh page to retry or try different CID

I see your approval, but I will work on that error message prior to merging.

**Comment by SgtPooki** [comment]
@lidel PR out to address this in ipld-explorer-components: https://github.com/ipfs/ipld-explorer-components/pull/378

**Comment by SgtPooki** [comment]
new preview url, https://bafybeif65hxysjwygngoahx7v6a2ouzsfi4y24u7ovr4vd5ya4jhq7ngjy.on.fleek.co/, updated in description. 

tested new error message at https://bafybeif65hxysjwygngoahx7v6a2ouzsfi4y24u7ovr4vd5ya4jhq7ngjy.on.fleek.co/#/explore/bafybeibms6xfhs4kx3tun7dg6bjbj5obca7yj43cl5jl3pc2n7muw6ipfm

image

**Comment by SgtPooki** [comment]
Changes deployed on explore.ipld.io.

https://explore.ipld.io/#/explore/bafybeibms6xfhs4kx3tun7dg6bjbj5obca7yj43cl5jl3pc2n7muw6ipfm shows the error message appropriately.

**Pull request merged**